client.go

```diff
@@ -84,47 +83,38 @@ func (codec *clientCodec) ReadResponseHeader(response *rpc.Response) (err error)
 
 	codec.mutex.Lock()
 	httpResponse := codec.responses[seq]
+	delete(codec.responses, seq)
 	codec.mutex.Unlock()
 
 	if httpResponse.StatusCode < 200 || httpResponse.StatusCode >= 300 {
-		return fmt.Errorf("request error: bad status code - %d", httpResponse.StatusCode)
+		response.Error = fmt.Sprintf("request error: bad status code - %d", httpResponse.StatusCode)
 	}
 
-	respData, err := ioutil.ReadAll(httpResponse.Body)
-
+	defer httpResponse.Body.Close()
+	body, err := ioutil.ReadAll(httpResponse.Body)
 	if err != nil {
-		return err
+		response.Error = err.Error()
 	}
 
-	httpResponse.Body.Close()
-
-	resp := NewResponse(respData)
-
-	if resp.Failed() {
-		response.Error = fmt.Sprintf("%v", resp.Err())
+	resp := Response(body)
+	if err := resp.Err(); err != nil {
+		response.Error = err.Error()
```

Comment:
If I see it correctly we might override error which is set in line 96 here. `ioutil.ReadAll` might return `nil` and an error. In this case we will call `resp.Err()` on an nil slice which will return unmarshalling error.